lib/rack/cache/meta_store.rb

```diff
@@ -64,6 +64,18 @@ def store(request, response, entity_store)
         end
         response.headers['X-Content-Digest'] = digest
         response.headers['Content-Length'] = size.to_s unless response.headers['Transfer-Encoding']
+
+        # If the entitystore backend is a Noop, do not try to read the body from the backend, it always returns an empty array
+        unless entity_store.is_a? Rack::Cache::EntityStore::Noop
+          # A stream body can only be read once and is currently closed by #write.
+          # (To avoid having to keep giant objects in memory when writing to disk cache
+          # the body is never converted to a single string)
+          # We cannot always reply on body to be re-readable,
+          # so we have to read it from the cache.
+          # BUG: if the cache was unable to store a stream, the stream will be closed
+          #      and rack will try to read it again, resulting in hard to track down exception
+          response.body = entity_store.open(digest) || response.body
+        end
```

Comment:
we have to support empty bodies too, for now I think the Noop check is the safest, I want to clean this up to make write return the written body so this re-reading goes away ...